build_tools/scripts/export_submodule_head.py

```diff
@@ -0,0 +1,225 @@
+        git_exec(
+            ["push", "origin", f"{submodule_head}:refs/heads/{submodule_branch}"],
+            repo_dir=submodule_path,
+        )
+        print("PLEASE IGNORE ANY NOTICE ABOUT CREATING A PR")
+        return 0
+
+    # Check if the submodule_head is an ancestor of the current remote_topic_head
+    # and exit if so (it is already reachable).
+    try:
+        git_exec(
+            ["merge-base", "--is-ancestor", submodule_head, remote_topic_head],
+            repo_dir=submodule_path,
+        )
+        print(
+            f"Commit {submodule_head} is reachable from remote branch {submodule_branch}. Doing nothing."
+        )
+        return
+    except subprocess.CalledProcessError as e:
+        # If not an ancestor, returncode will be 1. On general error, it will be
+        # something else.
+        if e.returncode != 1:
+            raise
+
+    # Create a splice commit that is based on the tree of the current submodule head
+    # and has parents of the current submodule head and the remote topic head.
+    # Note that the current branch is not touched, the commit is just created in the
+    # ether. We can push it to the remote topic branch to complete the splice.
+    print(f"Submodule head {submodule_head} is not on {submodule_branch}. Splicing.")
+    splice_commit = git_exec(
+        [
+            "commit-tree",
+            submodule_head + "^{tree}",
+            "-p",
+            submodule_head,
+            "-p",
+            remote_topic_head,
+            "-m",
+            f"Splice submodule rebase {submodule_head} onto {remote_topic_head}",
+        ],
+        repo_dir=submodule_path,
+        capture_output=True,
+    ).strip()
+    print(f"Created splice commit {splice_commit}: pushing")
+    git_exec(
+        ["push", "origin", f"{splice_commit}:refs/heads/{submodule_branch}"],
+        repo_dir=submodule_path,
+    )
+
+
+def git_current_commit(*, repo_dir=None) -> tuple[str, str]:
+    output = git_exec(
+        ["log", "-n", "1", "--pretty=format:%H %s (%an on %ci)"],
+        capture_output=True,
+        repo_dir=repo_dir,
+        quiet=True,
+    )
+    output = output.strip()
+    parts = output.split(" ")
+    # Return commit, full_summary
+    return parts[0], output
+
+
+def git_current_branch(*, repo_dir=None):
+    return git_exec(
+        ["rev-parse", "--abbrev-ref", "HEAD"],
+        repo_dir=repo_dir,
+        quiet=True,
+        capture_output=True,
+    ).strip()
+
+
+def check_origin_update_help(repo_dir):
+    existing_url = git_exec(
+        ["remote", "get-url", "--push", "origin"],
+        capture_output=True,
+        repo_dir=repo_dir,
+        quiet=True,
+    )
+    existing_url = existing_url.strip()
+    if existing_url.startswith("https://github.com/"):
+        new_url = existing_url.replace("https://github.com/", "git@github.com:", 1)
+        print(
+            "Your push URL is for GitHub HTTPS. You may need to switch to ssh for interactive use:"
```

Comment:
I locally removed this. HTTPS works with git tokens, so this is not true. I think we may need a better check for this.

Reply:
Turned it more into a helpful tip that doesn't block.